homeassistant/components/climate/generic_thermostat.py

```diff
@@ -231,33 +311,33 @@ def operation_list(self):
 
     async def async_set_operation_mode(self, operation_mode):
         """Set operation mode."""
-        if operation_mode == STATE_HEAT:
-            self._current_operation = STATE_HEAT
-            self._enabled = True
-            self._async_control_heating()
-        elif operation_mode == STATE_COOL:
-            self._current_operation = STATE_COOL
-            self._enabled = True
-            self._async_control_heating()
-        elif operation_mode == STATE_OFF:
-            self._current_operation = STATE_OFF
-            self._enabled = False
-            if self._is_device_active:
+        if operation_mode == STATE_OFF:
+            self._current_operation = operation_mode
+            if self._is_heating:
                 self._heater_turn_off()
+            if self._is_cooling:
+                self._ac_turn_off()
+        elif operation_mode in self.operation_list:
+            self._current_operation = operation_mode
+            self._async_control_thermostat()
         else:
             _LOGGER.error("Unrecognized operation mode: %s", operation_mode)
             return
-        # Ensure we update the current operation after changing the mode
         self.schedule_update_ha_state()
```

Comment:
This is calling a sync method from an async context. That is wrong. It should be `async_schedule_update_ha_state`.